### PR TITLE
Add base implementation for in-game leaderboards

### PR DESCRIPTION
**Pull request by frenzibyte**
Adds a base implementation of in-game leaderboards containing functions for adding real-time players (requires bindable double for current score) and adding a score.

### CODE REVIEW

File: osu.Game/Screens/Play/InGameLeaderboard.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Users;
+
+namespace osu.Game.Screens.Play
+{
+    public class InGameLeaderboard : CompositeDrawable
```
Comment (UselessToucan) [inline osu.Game/Screens/Play/InGameLeaderboard.cs]: `Leaderboard`

File: osu.Game/Screens/Play/InGameScoreContainer.cs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+using osu.Framework.Graphics.Containers;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Scoring;
+using osu.Game.Users;
+using osuTK;
+
+namespace osu.Game.Screens.Play
+{
+    public class InGameScoreContainer : FillFlowContainer<InGameScoreItem>
```
Comment (UselessToucan) [inline osu.Game/Screens/Play/InGameScoreContainer.cs]: `ScoreContainer`

File: osu.Game/Screens/Play/InGameScoreContainer.cs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            {
+                int newPosition = orderedPositions[i] ?? maxPosition + 1;
+
+                SetLayoutPosition(orderedByScore[i], newPosition);
+                orderedByScore[i].ScorePosition = DeclareNewPosition ? newPosition : orderedPositions[i];
+            }
+        }
+    }
+
+    public class InGameScoreItem : CompositeDrawable
```
Comment (UselessToucan) [inline osu.Game/Screens/Play/InGameScoreContainer.cs]: `ScoreDrawable`

File: osu.Game/Screens/Play/InGameLeaderboard.cs (hunk 1 above)
Comment (iiSaLMaN) [reply osu.Game/Screens/Play/InGameLeaderboard.cs]: will become confusing as hell with the leaderboard component, better to add an `InGame` before the components to mark them as in-game components

File: osu.Game/Screens/Play/InGameScoreContainer.cs (hunk 2 above)
Comment (iiSaLMaN) [reply osu.Game/Screens/Play/InGameScoreContainer.cs]: same as above, there is `PaginatedScoreContainer` for example

File: osu.Game/Screens/Play/InGameScoreContainer.cs (hunk 3 above)
Comment (iiSaLMaN) [reply osu.Game/Screens/Play/InGameScoreContainer.cs]: maybe `DrawableInGameScore`?

**Comment by yousef157** [comment]
Test

**Comment by peppy** [comment]
Is this supposed to use https://github.com/ppy/osu/pull/5416?

**Comment by iiSaLMaN** [comment]
Nope, `ILeaderboard` is required for reading the list of scores which will come out in a separate PR (retrieving scores from leaderboards), will continue working on that later.

**Comment by peppy** [comment]
Structurally this is in an OK state now. The bindable logic is a bit haphazard but I am working around that while integrating with the realtime multiplayer backend, so not touching it for now.

**Comment by peppy** [comment]
Going to merge this, pending further building on it in an upcoming PR (for multiplayer usage).

**Review by peppy — APPROVED** [review]

**Pull request merged**